File: app/src/main/java/de/euhm/jlt/listadapter/ViewTimesListAdapter.java
/*
 * @file ViewTimesListAdapter.java
 * @author Holger Mueller
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * http://www.apache.org/licenses/LICENSE-2.0
 */
package de.euhm.jlt.listadapter;

import java.util.Calendar;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import androidx.annotation.NonNull;

import de.euhm.jlt.dao.Times;
import de.euhm.jlt.utils.TimeUtil;
import de.euhm.jlt.R;

/**
 * List adapter for Times
 * @author hmueller
 * @version $Rev: 46 $
 */
public class ViewTimesListAdapter extends ArrayAdapter<Times> {
    private final String LOG_TAG = ViewTimesListAdapter.class.getSimpleName();

    private Context mContext;
    private List<Times> mListTimes;
	private int mLayoutId;
 
    public ViewTimesListAdapter(Context context, int layoutId, 
		List<Times> listTimes) {
        super(context, layoutId, listTimes);
        Log.d(LOG_TAG, "Creating list adapter");

        mContext = context;
		mLayoutId = layoutId;
        mListTimes = listTimes;
    }

    @NonNull
	@Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        //Log.d(LOG_TAG, "Start getView position " + position);

        if (convertView == null) {
            //Log.d(LOG_TAG, "Creating new line in the list");
            LayoutInflater inflater = (LayoutInflater)mContext.getSystemService
            	      (Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(mLayoutId, parent, false);
        }
		// If convertView is not null, we can just reuse it from the recycler
		
        Times times = mListTimes.get(position);
		Calendar timeStart = Calendar.getInstance();
		timeStart.setTimeInMillis(times.getTimeStart());
		Calendar timeEnd = Calendar.getInstance();
		timeEnd.setTimeInMillis(times.getTimeEnd());
		
        TextView dayOfMonth = (TextView) convertView.findViewById(R.id.view_times_day_of_month);
        TextView dayOfWeek = (TextView) convertView.findViewById(R.id.view_times_day_of_week);
        TextView monthAndYear = (TextView) convertView.findViewById(R.id.view_times_month_and_year);
        TextView hours = (TextView) convertView.findViewById(R.id.view_times_hours);
        TextView duration = (TextView) convertView.findViewById(R.id.view_times_duration);
		TextView homeOffice = (TextView) convertView.findViewById(R.id.view_times_homeoffice);

        dayOfMonth.setText(String.format(Locale.getDefault(), "%td", timeStart));
        dayOfWeek.setText(String.format(Locale.getDefault(), "%tA (%d)", timeStart, timeStart.get(Calendar.WEEK_OF_YEAR)));
        monthAndYear.setText(String.format(Locale.getDefault(), "%1$tB %1$tY", timeStart));
        hours.setText(String.format(Locale.getDefault(), "%tR - %tR", timeStart, timeEnd));
        duration.setText(TimeUtil.formatTimeString24(TimeUtil.getWorkedTime(mContext, timeStart, timeEnd, times.getHomeOffice())) + " (" +
        		TimeUtil.formatTimeString24(TimeUtil.getOverTime(mContext, timeStart, timeEnd, times.getHomeOffice())) + ")");
		homeOffice.setVisibility(times.getHomeOffice() ? View.VISIBLE : View.INVISIBLE);

        return convertView;
    }

    /**
     * Remove object based on list id.
     * 
     * If list is rebuild and object is from the list before, object.equals(a[i]) in class.
     * ArrayList fails. So do the compare with the object.getId().
     * 
     * Does <b>not</b> do a notifyDataSetChanged(), you have to do this manually!
     * 
     * @param object Times object 
     */
    @Override
    public void remove(Times object) {
    	if (object != null) {
    		int s = mListTimes.size();
    		for (int i = 0; i < s; i++) {
    			Times o = mListTimes.get(i);
    			if (o.getId() == object.getId()) {
    				mListTimes.remove(i);
    				return;
    			}
    		}
    	}
    }
    
    /**
     * Update object based on list id.
     * Does <b>not</b> do a notifyDataSetChanged(), you have to do this manually!
     * 
     * @param object Times object 
     */
    public void update(Times object) {
    	if (object != null) {
    		int s = mListTimes.size();
    		for (int i = 0; i < s; i++) {
    			Times o = mListTimes.get(i);
    			if (o.getId() == object.getId()) {
    				mListTimes.set(i, object);
    				return;
    			}
    		}
    	}
    }
    
    /**
     * Sort the list
     */
	public void sort() {
		sort(new Comparator<Times>() {
			@Override
			public int compare(Times a, Times b) {
				return a.getTimeStart() > b.getTimeStart() ? -1 : a.getTimeStart() == b.getTimeStart() ? 0 : 1;
			}
		});
	}

	/**
	 * Refill the list. Does a notifyDataSetChanged().
	 * @param listTimes List of new times values
	 */
    public void refill(List<Times> listTimes) {
        mListTimes.clear();
        mListTimes.addAll(listTimes);
        notifyDataSetChanged();
    }
}

File: app/src/main/java/de/euhm/jlt/listadapter/ViewTimesListAdapter.kt
/**
 * @file ViewTimesListAdapter.kt
 * @author Holger Mueller
 *
 * Converted from Java to Kotlin by Android Studio Meerkat | 2024.3.1 in March 2025
 *
 * MIT License
 * Copyright (c) 2014-2025 Holger Mueller
 */
package de.euhm.jlt.listadapter

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import de.euhm.jlt.R
import de.euhm.jlt.dao.Times
import de.euhm.jlt.utils.TimeUtil.formatTimeString24
import de.euhm.jlt.utils.TimeUtil.getOverTime
import de.euhm.jlt.utils.TimeUtil.getWorkedTime
import java.util.Calendar
import java.util.Locale

/**
 * List adapter for Times
 *
 * @author hmueller
 */
class ViewTimesListAdapter(context: Context, layoutId: Int, listTimes: List<Times>) :
    ArrayAdapter<Times>(context, layoutId, listTimes) {
    @Suppress("PrivatePropertyName")
    private val LOG_TAG: String = ViewTimesListAdapter::class.java.simpleName

    private val mContext: Context
    private val mListTimes: MutableList<Times>
    private val mLayoutId: Int

    init {
        Log.d(LOG_TAG, "Creating list adapter")

        mContext = context
        mLayoutId = layoutId
        mListTimes = listTimes as MutableList<Times>
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        //Log.d(LOG_TAG, "Start getView position " + position);

        val inflater = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        // If convertView is not null, we can just reuse it from the recycler
        val cvtView = convertView ?: inflater.inflate(mLayoutId, parent, false)

        val times = mListTimes[position]
        val timeStart = Calendar.getInstance()
        timeStart.timeInMillis = times.timeStart
        val timeEnd = Calendar.getInstance()
        timeEnd.timeInMillis = times.timeEnd

        val dayOfMonth = cvtView.findViewById<TextView>(R.id.view_times_day_of_month)
        val dayOfWeek = cvtView.findViewById<TextView>(R.id.view_times_day_of_week)
        val monthAndYear = cvtView.findViewById<TextView>(R.id.view_times_month_and_year)
        val hours = cvtView.findViewById<TextView>(R.id.view_times_hours)
        val duration = cvtView.findViewById<TextView>(R.id.view_times_duration)
        val homeOffice = cvtView.findViewById<TextView>(R.id.view_times_homeoffice)

        dayOfMonth.text = String.format(Locale.getDefault(), "%td", timeStart)
        dayOfWeek.text = String.format(Locale.getDefault(), "%tA (%d)", timeStart, timeStart[Calendar.WEEK_OF_YEAR])
        monthAndYear.text = String.format(Locale.getDefault(), "%1\$tB %1\$tY", timeStart)
        hours.text = String.format(Locale.getDefault(), "%tR - %tR", timeStart, timeEnd)
        duration.text = String.format(Locale.getDefault(),
            "%s (%s)",
            formatTimeString24(getWorkedTime(mContext, timeStart, timeEnd, times.homeOffice)),
            formatTimeString24(getOverTime(mContext, timeStart, timeEnd, times.homeOffice)))
        homeOffice.visibility = if (times.homeOffice) View.VISIBLE else View.INVISIBLE

        return cvtView
    }

    /**
     * Remove object based on list id.
     *
     * If list is rebuild and object is from the list before, object.equals(a[i ]) in class.
     * ArrayList fails. So do the compare with the object.getId().
     *
     * Does **not** do a notifyDataSetChanged(), you have to do this manually!
     *
     * @param obj Times object
     */
    override
    fun remove(obj: Times?) {
        obj?.let {  // code inside the let block will only execute if obj is not null
            val index = mListTimes.indexOfFirst { it.id == obj.id }
            if (index != -1) {
                mListTimes.removeAt(index)
            }
        }
    }

    /**
     * Update object based on list id.
     * Does **not** do a notifyDataSetChanged(), you have to do this manually!
     *
     * @param obj Times object
     */
    fun update(obj: Times?) {
        obj?.let {  // code inside the let block will only execute if obj is not null
            val index = mListTimes.indexOfFirst { it.id == obj.id }
            if (index != -1) {
                mListTimes[index] = obj
            }
        }
    }


    /**
     * Sort the list
     */
    fun sort() {
        sort { a: Times, b: Times -> b.timeStart.compareTo(a.timeStart) }
    }

    /**
     * Refill the list. Does a notifyDataSetChanged().
     *
     * @param listTimes List of new times values
     */
    fun refill(listTimes: List<Times>) {
        mListTimes.clear()
        mListTimes.addAll(listTimes)
        notifyDataSetChanged()
    }
}